### PR TITLE
restrict access to /admin to whitelisted ips

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

We want to give access to /admin path to a restricted list of ips we
want to control. If the whitelist is empty the restriction is disabled and if ips are added to this list they will be used to restrict the access of /admin to this given list.

## Proposal

- [x] configure nginx on edxapp
- [x] configure nginx on richie
- [x] configure nginx on marsha

### CODE REVIEW

File: apps/edxapp/templates/services/nginx/configs/lms.conf.j2 [hunk 1]
```diff
@@ -44,6 +44,17 @@ server {
     try_files $uri @proxy_to_lms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+
+    if ($http_x_forwarded_for !~ ({{  edxapp_nginx_admin_ip_withelist | join("|") }})) {
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/lms.conf.j2]:
```suggestion
    if ($http_x_forwarded_for !~ ({{ edxapp_nginx_admin_ip_withelist | join("|") }})) {
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
 
+### Added
+
+- Whitelist ips authorized to acess /admin on edxapp, marsha and richie
```
Comment (sampaccoud) [inline CHANGELOG.md]:
```suggestion
- Whitelist ips authorized to access /admin on edxapp, marsha and richie
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
...
+
```
Comment (sampaccoud) [inline CHANGELOG.md:16]: Would be nice to apply it also to edxec's admin pages and to all pages of mailcatcher and learninglocker?

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 4]
```diff
@@ -34,6 +34,17 @@ server {
     try_files $uri @proxy_to_cms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+
+    if ($http_x_forwarded_for !~ ({{ edxapp_nginx_admin_ip_withelist | join("|") }})) {
```
Comment (sampaccoud) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: I think `http_x_forwarded_for` is only set behind a proxy.

File: CHANGELOG.md (hunk 3 above)
Comment (lunika) [reply CHANGELOG.md]: You mean only if there is a haproxy/nginx/whatever in front of openshift ?

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 (hunk 4 above)
Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: You mean only if there is a haproxy/nginx/whatever in front of openshift ?

Comment (sampaccoud) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: From what I understand, `HTTP_X_FORWARDED_FOR` can be a list of IPs when you are behind a number of proxies. If not behind a proxy, you should probably use `HTTP_X_REAL_IP`.

Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: But is it possible to directly expose an openshift or kubernetes cluster ? I think it's always behind a proxy

Comment (sampaccoud) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: yes but even behind a proxy, `http_x_forwarded_for` is a list that grows and anybody can send whatever they want into it. We can use `http_x_real_ip` and specify that the openshift should be placed behind a proxy that is configured to set this header with the value of the real `remote_ip`?

Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: We have to use `http_x_forwarded_for` but keep in my mind that it's a comma separated list and the first occurence in the list is the real ip : https://github.com/symfony/http-foundation/blob/897bf05a4dd936cfd028d7d04f0f21590c024030/Request.php#L791-L798

Comment (sampaccoud) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: ok but if there is no proxy, http_x_forwarded_for is empty so you must default to `remote_ip`.

Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: I made a modification. We know that the real ip is always the first one in the `http_x_forwarded_for` variable so we check if the string starts with the ip we want. WDYT ?

File: CHANGELOG.md (hunk 3 above)
Comment (lunika) [reply CHANGELOG.md:16]: done for learninglocker but for mailcatcher there is no nginx. I will check if the application allow to do that.

File: apps/learninglocker/templates/services/nginx/configs/default.conf.j2 [hunk 5]
```diff
@@ -42,6 +42,16 @@ server {
 
   # All other traffic directed to statics or Node server
   location / {
+{% if learninglocker_nginx_admin_ip_withelist | length > 0 %}
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/learninglocker/templates/services/nginx/configs/default.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: apps/richie/templates/services/nginx/configs/richie.conf.j2 [hunk 6]
```diff
@@ -29,6 +29,20 @@ server {
     try_files $uri @proxy_to_richie_app;
   }
 
+{% if richie_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/richie/templates/services/nginx/configs/richie.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 7]
```diff
@@ -34,6 +34,20 @@ server {
     try_files $uri @proxy_to_cms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: apps/edxapp/templates/services/nginx/configs/lms.conf.j2 [hunk 8]
```diff
@@ -44,6 +44,20 @@ server {
     try_files $uri @proxy_to_lms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/edxapp/templates/services/nginx/configs/lms.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: apps/edxec/templates/services/nginx/configs/edxec.conf.j2 [hunk 9]
```diff
@@ -29,6 +29,20 @@ server {
     try_files $uri @proxy_to_edxec_app;
   }
 
+{% if edxec_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/edxec/templates/services/nginx/configs/edxec.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: apps/marsha/templates/services/nginx/configs/marsha.conf.j2 [hunk 10]
```diff
@@ -29,6 +29,20 @@ server {
     try_files $uri @proxy_to_marsha_app;
   }
 
+{% if marsha_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to check the first ip present in $http_x_forwarded_for, the other one are added
```
Comment (sampaccoud) [inline apps/marsha/templates/services/nginx/configs/marsha.conf.j2]:
```suggestion
      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -8,6 +8,12 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
 
+### Added
+
+- Whitelist ips authorized to access /admin on edxapp,
+  edxec, marsha and richie
+- Whitelist ips authorized to access to learninglocker's UI
```
Comment (sampaccoud) [inline CHANGELOG.md]:
```suggestion
- Whitelist ips authorized to access learninglocker's UI
```

**Review by sampaccoud — APPROVED** [review]

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 12]
```diff
@@ -34,6 +34,20 @@ server {
     try_files $uri @proxy_to_cms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
+  location /admin {
+    {# 
+      We want to limit access to a list of whitelisted IP addresses. Check the first IP present in 
+      $http_x_forwarded_for, the other one are added by proxies and are not interesting here.
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: I suggest the following comment (I don't understand the current one):

>We want to limit access to a list of whitelisted IP addresses.
>
>`$http_x_forwarded_for` variable contains a list of IPs listed from the `HTTP_X_FORWARED_FOR` header (_e.g._ `w.x.y.z, 10.0.0.1`). The first IP corresponds to the client's public address, which is of interest (other ones have been added by subsequent proxies), hence we restrict our comparison with the beginning of this list (this is why our regex starts with a `^`).

File: apps/edxapp/templates/services/nginx/configs/lms.conf.j2 [hunk 13]
```diff
@@ -44,6 +44,20 @@ server {
     try_files $uri @proxy_to_lms_app;
   }
 
+{% if edxapp_nginx_admin_ip_withelist | length > 0 %}
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/lms.conf.j2]: Should we define two separated lists for the CMS and the LMS ?

Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/lms.conf.j2]: Yes I will do that

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 14]
```diff
@@ -34,6 +34,24 @@ server {
     try_files $uri @proxy_to_cms_app;
   }
 
+{% if edxapp_nginx_cms_admin_ip_withelist | length > 0 %}
+  location /admin {
+        {#
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: weird indentation

Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: good catch

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**